keystore/keystore.go

```diff
@@ -0,0 +1,108 @@
+package keystore
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+)
+
+type CertificateInformation struct {
+	FirstAndLastName   string
+	OrganizationalUnit string
+	Organization       string
+	CityOrLocality     string
+	StateOrProvince    string
+	CountryCode        string
+	ValidFrom          string
+	ValidUntil         string
+}
+
+var (
+	IncorrectAliasError            = errors.New("incorrect key alias")
+	IncorrectKeystorePasswordError = errors.New("incorrect keystore password")
+	IncorrectKeyPasswordError      = errors.New("incorrect key password")
+)
+
+type Decoder interface {
+	Decode(data []byte, password, alias, keyPassword string) (privateKey interface{}, certificate *x509.Certificate, err error)
+	IsInvalidCredentialsError(err error) bool
+}
+
+type Reader struct {
+	decoders []Decoder
+}
+
+func NewReader(decoders []Decoder) Reader {
+	return Reader{decoders: decoders}
+}
+
+func NewDefaultReader() Reader {
+	return NewReader([]Decoder{PKCS12KeystoreDecoder{}, JKSKeystoreDecoder{}})
+}
+
+func (p Reader) ReadCertificateInformation(data []byte, password, alias, keyPassword string) (*CertificateInformation, error) {
+	var cert *x509.Certificate
+	var decodeErrs []error
+
+	for _, decoder := range p.decoders {
```

Comment:
It is perfectly fine to iterate over the decoders and try which one decodes the byte array. But do not the pkcs12 and jks format have some kind of distinct way to identify them. Just checking if instead of trying all of the decoders we could just simply initialise the decoder on the fly (knowing what kind of keystore format the byte array is in) and simplify this piece of code.

Reply:
I was thinking about it. I couldn't find an out-of-the-box solution for deciding which decoder should be used.
Probably, I could probably unwarp parts of the decoders and try to implement type detection, but because I'm not experienced in cryptography I would rather skip this.